File: frontend/src/app/features/admin/services/user-management.service.ts
import { Injectable, inject } from '@angular/core';
import { HttpClient, HttpParams } from '@angular/common/http';
import { Observable } from 'rxjs';
import { environment } from '@env/environment';
import { 
  UserManagement, 
  UpdateUserRoleRequest, 
  UpdateUserStatusRequest 
} from '../models/user-management.model';

/**
 * Service for admin user management operations
 */
@Injectable({
  providedIn: 'root'
})
export class UserManagementService {
  private http = inject(HttpClient);
  private apiUrl = `${environment.apiUrl}/admin/users`;

  /**
   * List all users with pagination
   */
  listUsers(page: number = 0, size: number = 50): Observable<UserManagement[]> {
    const params = new HttpParams()
      .set('page', page.toString())
      .set('size', size.toString());
    
    return this.http.get<UserManagement[]>(this.apiUrl, { params });
  }

  /**
   * Update user role (SUPERADMIN only)
   */
  updateUserRole(userId: string, request: UpdateUserRoleRequest): Observable<UserManagement> {
    return this.http.put<UserManagement>(`${this.apiUrl}/${userId}/role`, request);
  }

  /**
   * Enable/disable user account (ADMIN+)
   */
  updateUserStatus(userId: string, request: UpdateUserStatusRequest): Observable<UserManagement> {
    return this.http.put<UserManagement>(`${this.apiUrl}/${userId}/status`, request);
  }

  /**
   * Delete user (SUPERADMIN only)
   */
  deleteUser(userId: string): Observable<void> {
    return this.http.delete<void>(`${this.apiUrl}/${userId}`);
  }
}

File: backend/src/main/java/com/learnsinhala/service/UserManagementService.java
package com.learnsinhala.service;

import com.learnsinhala.dto.user.UpdateUserRoleRequest;
import com.learnsinhala.dto.user.UpdateUserStatusRequest;
import com.learnsinhala.dto.user.UserManagementDto;
import com.learnsinhala.exception.ApiException;
import com.learnsinhala.model.Role;
import com.learnsinhala.model.User;
import com.learnsinhala.repository.SentencePatternRepository;
import com.learnsinhala.repository.UserRepository;
import com.learnsinhala.repository.VocabularyRepository;
import com.learnsinhala.security.RoleValidator;
import lombok.RequiredArgsConstructor;
import lombok.extern.slf4j.Slf4j;
import org.springframework.data.domain.Page;
import org.springframework.data.domain.PageRequest;
import org.springframework.data.domain.Pageable;
import org.springframework.data.domain.Sort;
import org.springframework.stereotype.Service;

import java.util.List;
import java.util.stream.Collectors;

/**
 * Service for user management operations (ADMIN+).
 */
@Service
@RequiredArgsConstructor
@Slf4j
public class UserManagementService {

    private final UserRepository userRepository;
    private final VocabularyRepository vocabularyRepository;
    private final SentencePatternRepository sentencePatternRepository;

    /**
     * List all users with pagination.
     * Only ADMIN and SUPERADMIN can access.
     */
    public List<UserManagementDto> listUsers(User requester, Integer page, Integer size) {
        // Check permission
        if (!RoleValidator.canManageUsers(requester)) {
            throw ApiException.forbidden("Only ADMIN and SUPERADMIN can list users");
        }

        Pageable pageable = PageRequest.of(
                page != null ? page : 0,
                size != null ? size : 50,
                Sort.by(Sort.Direction.DESC, "createdAt")
        );

        Page<User> usersPage = userRepository.findAll(pageable);

        return usersPage.getContent().stream()
                .map(user -> {
                    // Get content counts for each user
                    long vocabCount = vocabularyRepository.countByCreatedBy(user.getId());
                    long patternCount = sentencePatternRepository.countByCreatedBy(user.getId());
                    return UserManagementDto.from(user, vocabCount, patternCount);
                })
                .collect(Collectors.toList());
    }

    /**
     * Update user role.
     * Only SUPERADMIN can perform this operation.
     */
    public UserManagementDto updateUserRole(User requester, String userId, UpdateUserRoleRequest request) {
        // Check permission - only SUPERADMIN can change roles
        if (!RoleValidator.isSuperAdmin(requester)) {
            throw ApiException.forbidden("Only SUPERADMIN can update user roles");
        }

        // Cannot modify your own role
        if (requester.getId().equals(userId)) {
            throw ApiException.badRequest("Cannot modify your own role");
        }

        User targetUser = userRepository.findById(userId)
                .orElseThrow(() -> ApiException.notFound("User not found"));

        targetUser.setRole(request.getRole());
        targetUser = userRepository.save(targetUser);

        log.info("SUPERADMIN {} updated user {} role to {}", requester.getEmail(), userId, request.getRole());

        return UserManagementDto.from(targetUser);
    }

    /**
     * Enable or disable user account.
     * ADMIN and SUPERADMIN can perform this operation.
     */
    public UserManagementDto updateUserStatus(User requester, String userId, UpdateUserStatusRequest request) {
        // Check permission
        if (!RoleValidator.canManageUsers(requester)) {
            throw ApiException.forbidden("Only ADMIN and SUPERADMIN can update user status");
        }

        // Cannot modify your own status
        if (requester.getId().equals(userId)) {
            throw ApiException.badRequest("Cannot modify your own status");
        }

        User targetUser = userRepository.findById(userId)
                .orElseThrow(() -> ApiException.notFound("User not found"));

        targetUser.setEnabled(request.getEnabled());
        targetUser = userRepository.save(targetUser);

        log.info("{} {} user {}", requester.getEmail(), 
                request.getEnabled() ? "enabled" : "disabled", userId);

        return UserManagementDto.from(targetUser);
    }

    /**
     * Delete user.
     * Only SUPERADMIN can perform this operation.
     */
    public void deleteUser(User requester, String userId) {
        // Check permission - only SUPERADMIN can delete users
        if (!RoleValidator.isSuperAdmin(requester)) {
            throw ApiException.forbidden("Only SUPERADMIN can delete users");
        }

        // Cannot delete yourself
        if (requester.getId().equals(userId)) {
            throw ApiException.badRequest("Cannot delete your own account");
        }

        User targetUser = userRepository.findById(userId)
                .orElseThrow(() -> ApiException.notFound("User not found"));

        // Delete the user
        // Note: Content created by this user will remain (orphaned with their userId in createdBy)
        userRepository.deleteById(userId);

        log.warn("SUPERADMIN {} deleted user {}", requester.getEmail(), targetUser.getEmail());
    }
}
